spring-scheduling/src/test/java/com/baeldung/asyncwithretry/EventServiceIntegrationTest.java

```diff
@@ -0,0 +1,90 @@
+package com.baeldung.asyncwithretry;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.mock.mockito.MockBean;
+import org.springframework.test.context.ContextConfiguration;
+import org.springframework.test.context.junit.jupiter.SpringExtension;
+import org.springframework.test.context.support.AnnotationConfigContextLoader;
+import org.springframework.web.client.HttpServerErrorException;
+
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+
+import static org.junit.jupiter.api.Assertions.*;
+import static org.mockito.ArgumentMatchers.anyList;
+import static org.mockito.Mockito.*;
+
+@ExtendWith(SpringExtension.class)
+@ContextConfiguration(classes = {AsyncConfig.class}, loader = AnnotationConfigContextLoader.class)
+class EventServiceIntegrationTest {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(EventServiceIntegrationTest.class);
+
+    @Autowired
+    private EventService asyncEventService;
+
+    @MockBean
+    private DownstreamService downstreamService;
+
+    @Test
+    void givenAsyncMethodHasNoException_whenAsyncMethodIscalled_thenReturnSuccess_WithoutAnyRetry() throws InterruptedException, ExecutionException {
+        LOGGER.info("Testing for async with retry execution with thread " + Thread.currentThread().getName());
+        when(downstreamService.publishEvents(anyList())).thenReturn(true);
+        Future<String> resultFuture = asyncEventService.processEvents(List.of("test1"));
+
+        while (!resultFuture.isDone() && !resultFuture.isCancelled()) {
+        }
```

Comment:
I suggest putting a `Thread.sleep(5)` in this loop, otherwise the test's own thread will be whizzing round burning up CPU units and slowing down the system. (in all of these `whiles`)

Reply:
Updated that with the more readable version of Thread.sleep() -> TimeUnit.SECONDS.sleep() which does the same thing.